ably-ios/ARTEventEmitter.m

```diff
@@ -6,39 +6,201 @@
+    return listener;
 }
 
-- (void)removeEvents {
-    [self.realtime.stateSubscriptions removeAllObjects];
+- (ARTEventListener *)once:(id)event call:(void (^)(id __art_nonnull))cb {
+    ARTEventListener *listener = [[ARTEventListener alloc] initWithBlock:cb];
+    [self addOnEntry:[[ARTEventEmitterEntry alloc] initWithListener:listener once:true] event:event];
+    return listener;
+}
+
+- (void)addOnEntry:(ARTEventEmitterEntry *)entry event:(id)event {
+    [self addObject:entry toArrayWithKey:event inDictionary:self.listeners];
+}
+
+- (ARTEventListener *)on:(void (^)(id __art_nonnull))cb {
+    ARTEventListener *listener = [[ARTEventListener alloc] initWithBlock:cb];
+    [self addOnAllEntry:[[ARTEventEmitterEntry alloc] initWithListener:listener once:false]];
+    return listener;
+}
+
+- (ARTEventListener *)once:(void (^)(id __art_nonnull))cb {
+    ARTEventListener *listener = [[ARTEventListener alloc] initWithBlock:cb];
+    [self addOnAllEntry:[[ARTEventEmitterEntry alloc] initWithListener:listener once:true]];
+    return listener;
+}
+
+- (void)addOnAllEntry:(ARTEventEmitterEntry *)entry {
+    [self.anyListeners addObject:entry];
+}
+
+- (void)off:(id)event listener:(ARTEventListener *)listener {
+    [self removeObject:listener fromArrayWithKey:event inDictionary:self.listeners where:^BOOL(id entry) {
+        return ((ARTEventEmitterEntry *)entry).listener == listener;
+    }];
+}
+
+- (void)off:(ARTEventListener *)listener {
+    BOOL (^cond)(id) = ^BOOL(id entry) {
+        return ((ARTEventEmitterEntry *)entry).listener == listener;
+    };
+    [self.anyListeners artRemoveWhere:cond];
+    for (id event in [self.listeners allKeys]) {
+        [self removeObject:listener fromArrayWithKey:event inDictionary:self.listeners where:cond];
+    }
+}
+
+- (void)off {
+    [self resetListeners];
+}
+
+- (void)resetListeners {
+    _listeners = [[NSMutableDictionary alloc] init];
+    _anyListeners = [[NSMutableArray alloc] init];
+}
+
+- (void)emit:(id)event with:(id)data {
+    NSMutableArray *listenersForEvent = [self.listeners objectForKey:event];
+    NSMutableArray *toCall = [[NSMutableArray alloc] init];
+    NSMutableArray *toRemoveFromListenersForEvent = [[NSMutableArray alloc] init];
+    NSMutableArray *toRemoveFromTotalListeners = [[NSMutableArray alloc] init];
+    @try {
+        for (ARTEventEmitterEntry *entry in listenersForEvent) {
+            if (entry.once) {
+                [toRemoveFromListenersForEvent addObject:entry];
+            }
+            
+            [toCall addObject:entry];
+        }
+        
+        for (ARTEventEmitterEntry *entry in self.anyListeners) {
+            if (entry.once) {
+                [toRemoveFromTotalListeners addObject:entry];
+            }
+            [toCall addObject:entry];
+        }
+    }
+    @finally {
+        for (ARTEventEmitterEntry *entry in toRemoveFromListenersForEvent) {
+            [listenersForEvent removeObject:entry];
```

Comment:
Hey, good catch. Although I'm not sure what you mean. This should be calling `[removeObject:entry fromArrayWithKey:event inDictionary:self.listeners]`, because that handles the case in which `self.listeners[event]` is left empty and needs to be deleted to avoid entry leaks. I fixed that at 859e474.